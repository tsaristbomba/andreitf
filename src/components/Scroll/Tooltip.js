import React from "react"
import styled from "styled-components"
import PropTypes from "prop-types"

const TooltipWrapper = styled.div`
  display: none;

  @media screen and (min-width: 768px) {
    z-index: 12;
    position: fixed;
    /* bottom: 7vh; */
    opacity: 1;
    right: 13.8vh;
    background: ${({ colors }) => colors.primary};
    color: ${({ colors }) => colors.tertiary};
    padding: 0 10px;
    display: inline-flex;
    max-width: 150px;
    transition: 0.2s opacity ease;
    font-size: 12px;
    border-radius: ${({ rounded }) => (rounded ? "4px" : "0")};
    animation: 0.3s fade;

    /* To the right of the tooltip */
    ::after {
      content: " ";
      position: absolute;
      top: 50%;
      left: 100%;
      margin-top: -5px;
      border-width: 5px;
      border-style: solid;
      border-color: transparent transparent transparent
        ${({ colors }) => colors.primary};
    }
  }
`

const Tooltip = ({ children, rounded, colors }) => {
  return (
    <TooltipWrapper colors={colors} rounded={rounded}>
      {children}
    </TooltipWrapper>
  )
}

Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  colors: PropTypes.object,
  rounded: PropTypes.bool,
}

export default Tooltip
