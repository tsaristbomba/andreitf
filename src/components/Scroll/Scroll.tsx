import React, { useState, useEffect } from "react"
import styled from "styled-components"
import { BsArrowUp } from "@react-icons/all-files/bs/BsArrowUp"

// Components
import Tooltip from "./Tooltip"

// Utils
import getMetadata from "../../utils/getMetadata"
import handleHexToRgba from "../../utils/handleHexToRgba"

// Types
type ScrollButtonTypes = {
  rounded?: boolean
  data: { message: string }
  msg?: boolean
}
type ScrollButtonStyledTypes = {
  rounded?: boolean
  colors: {
    primary?: string
    secondary?: string
    tertiary?: string
  }
  scrolled?: boolean
}
type MetadataColorsTypes = {
  primary?: string
  secondary?: string
  tertiary?: string
}

// Styled components
const Button = styled.button`
  display: ${(props: ScrollButtonStyledTypes) =>
    props.scrolled ? "flex" : "none"};
  z-index: 12;
  position: fixed;
  bottom: 7.3vh;
  right: 4.3vh;
  height: 30px;
  width: 30px;
  cursor: pointer;
  background: ${({ colors }) => colors.primary};
  border: none;
  border-radius: ${({ rounded }) => (rounded ? "50%" : "none")};
  align-items: center;
  justify-content: center;
  box-shadow: 1px 1px 3px 3px
    ${props => handleHexToRgba(props.colors.primary, 0.2)};
  transition: background 0.2s ease;

  // Transition group
  .transition-enter {
    opacity: 0.01;
    transform: translate(0, -10px);
  }
  .transition-enter-active {
    opacity: 1;
    transform: translate(0, 0);
    transition: opacity 300ms ease-in, transform 300ms ease-in;
  }
  .transition-exit {
    opacity: 1;
    transform: translate(0, 0);
  }
  .transition-exit-active {
    opacity: 0.01;
    transform: translate(0, 10px);
    transition: opacity 300ms ease-in, transform 300ms ease-in;
  }
  //

  svg {
    color: ${({ colors }) => colors.tertiary};
  }

  :hover {
    background: ${({ colors }) => colors.secondary};
    color: ${({ colors }) => colors.primary};
  }

  @media screen and (min-width: 768px) {
    right: 8.2vh;
    height: 30px;
    width: 30px;
  }
`
const Arrow = styled(BsArrowUp)`
  font-size: 1.5rem;
  color: ${(props: ScrollButtonStyledTypes) => props.colors.tertiary};
`

const ScrollButton: React.FC<ScrollButtonTypes> = ({
  rounded,
  data,
  msg,
}): JSX.Element => {
  const [scrolled, setScrolled] = useState<boolean>(false)
  const [tooltip, setTooltip] = React.useState<boolean>(true)

  const { primary, secondary, tertiary }: MetadataColorsTypes =
    getMetadata().site.siteMetadata.colors

  function scrollToTop() {
    window.scrollTo(0, 0)
  }

  function handleScroll() {
    if (window.scrollY >= 200) {
      setScrolled(true)
    } else {
      setScrolled(false)
    }
  }

  React.useEffect(() => {
    // Deactivate Tooltip after 10s
    if (msg) {
      setTimeout(() => {
        setTooltip(false)
        // setIsOpen(false);
      }, 5000)
    } else {
      setTooltip(false)
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  useEffect(() => {
    window.addEventListener("scroll", handleScroll)

    return function cleanup() {
      window.removeEventListener("scroll", handleScroll)
    }
  }, [])

  return (
    <Button
      onMouseEnter={() => setTooltip(true)}
      onMouseLeave={() => setTooltip(false)}
      rounded={rounded}
      scrolled={scrolled}
      onClick={scrollToTop}
      aria-label={data.message}
      colors={{ primary, tertiary, secondary }}
    >
      <Arrow colors={{ tertiary }} />
      {tooltip && (
        <Tooltip colors={{ primary, tertiary }} rounded={rounded}>
          {data.message}
        </Tooltip>
      )}
    </Button>
  )
}

export default ScrollButton
