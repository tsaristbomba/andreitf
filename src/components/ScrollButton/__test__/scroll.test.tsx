import React from "react"
import renderer from "react-test-renderer"
import Scroll from "../index"

describe("Scroll", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Scroll />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
