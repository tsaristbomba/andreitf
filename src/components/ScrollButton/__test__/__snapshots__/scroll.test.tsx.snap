// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll renders correctly 1`] = `
<button
  className="sc-bdnxRM bhiJDY"
  onClick={[Function]}
>
  <svg
    className="sc-gtsrHT fBKbYI"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M8 3.5a.5.5 0 01.5.5v9a.5.5 0 01-1 0V4a.5.5 0 01.5-.5z"
      fillRule="evenodd"
    />
    <path
      clipRule="evenodd"
      d="M7.646 2.646a.5.5 0 01.708 0l3 3a.5.5 0 01-.708.708L8 3.707 5.354 6.354a.5.5 0 11-.708-.708l3-3z"
      fillRule="evenodd"
    />
  </svg>
</button>
`;
